File: April 30 Day Challenge/Week 3 April 15th - April 21st/Product of Array Except Self.cpp
/*
Given an array nums of n integers where n > 1,  return an array output such that output[i] is equal to the product of all the elements of nums except nums[i].

Example:

Input:  [1,2,3,4]
Output: [24,12,8,6]

Constraint: It's guaranteed that the product of the elements of any prefix or suffix of the array (including the whole array) fits in a 32 bit integer.

Note: Please solve it without division and in O(n).

Follow up:
Could you solve it with constant space complexity? (The output array does not count as extra space for the purpose of space complexity analysis.)
*/

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        // int p = totalProduct(nums);
        vector<int> sumleft(nums.size());
        sumleft[0] = 1;
        
        for(int i=1; i<nums.size();i++){
            sumleft[i] = nums[i-1]*sumleft[i-1];
            // cout<<sumleft[i]<<" ";
        }
        
        int sumright = 1;
        int n = nums.size();
        for(int i=n-1; i>=0; i--){
            sumleft[i] *= sumright; 
            sumright *= nums[i];
           // cout<<sumleft[i]<<" ";
        }
        
        return sumleft;
        
    }
    
//     int totalProduct(vector<int> nums){
        
//         int ans = 1;
//         int n = nums.size();
//         for(int x: nums){
//             ans *= x;
//         }
        
//         return ans;
        
//     }
    
//     int divide(int a, int b){
//         long n = a, m = b;
//         int sign = (n < 0 ^ m < 0)? -1:1;
//         long long res = 0;
//         n = abs(n);
//         m = abs(m);
        
//         for(long t = 0, i = 31; i>=0 ; --i){
//             if(t + (m<<i) <= n){
//                 t += m<<i, res |= 1LL << i;   
//             }
            
//             // cout<<t<<" "<<res<<" ";
//         }
        
//         if(sign < 0){
//             res = -res; 
//         }
        
//         return (res >= INT_MAX || res < INT_MIN)? 0:res;
//     }
    
};

File: Medium/Product of Array Except Self.java
/*
Given an integer array nums, 
return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

You must write an algorithm that runs in O(n) time and without using the division operation.

 

Example 1:

Input: nums = [1,2,3,4]
Output: [24,12,8,6]
Example 2:

Input: nums = [-1,1,0,-3,3]
Output: [0,0,9,0,0]
 

Constraints:

2 <= nums.length <= 10^5
-30 <= nums[i] <= 30
The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
 

Follow up: Can you solve the problem in O(1) extra space complexity? 
(The output array does not count as extra space for space complexity analysis.)
*/

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int N = nums.length;
        int[] leftProduct= new int[N];
        leftProduct[0] = 1;
        
        for(int index = 1; index < N; ++index) {
            leftProduct[index] = leftProduct[index - 1]*nums[index - 1];
        }
        int rightProduct = 1;
        for(int index = N - 1; index >= 0; --index) {
            leftProduct[index] *= rightProduct;
            rightProduct *= nums[index];
        }
        
        return leftProduct;
    }
}